modules/shared/interpreter/src/main/scala/almond/interpreter/messagehandlers/InterpreterMessageHandlers.scala

```diff
@@ -168,6 +170,28 @@ final case class InterpreterMessageHandlers(
       inspectHandler
     )
 
+  def shutdownHandler: MessageHandler = {
+    // v5.3 spec states "The request can be sent on either the control or shell channels.".
+    def forChannel(c: Channel): MessageHandler = {
+      MessageHandler(c, Shutdown.requestType) { message =>
+        val reply = IO {
+          val reply = message.reply(Shutdown.replyType, Shutdown.Reply(message.content.restart))
+          c -> reply.asRawMessage
+        }
+
+        // Allow some time for message to be delivered. Doing this on the `queueEc` isn't great in that shutdown could
+        // hypothetically be sequenced behind queue operations, but in practice there's likely no consequence.
+        // Alternatively we could Thread.sleep() sleep inside a regular/synchronous IO but then might want to construct
+        // our Stream Chunks by hand to avoid sleep() causing any delay in delivery of our reply.
+        val delayedShutdown = IO.timer(queueEc).sleep(500.milliseconds).flatMap(_ => interpreter.shutdown)
```

Comment:
For later I guess, that could be fixed by managing not to trigger a call to `sys.exit` from the handler, but instead having the stream processing stuff cleanly return.

For example, one could create a `fs2.async.signalOf` in `almond.kernel.Kernel.replies`, pass it to the `shutdownHandler`, and have the handler set the signal to true.

From `almond.kernel.Kernel.replies`, `requests` would be read only until the signal is true (via the `interruptWhen` method of streams). That way, the processing of messages would cleanly stop after a shutdown request is received, and once all the messages are processed, and kernels would cleanly exit after that (the `unsafeRunSync()` [here](https://github.com/jupyter-scala/jupyter-scala/blob/46aa31ce9f11bc40bd6d92eb6112555e47da4354/modules/scala/scala-kernel/src/main/scala/almond/ScalaKernel.scala#L159) would just return, and the program would normally end).

I haven't actually tried that yet, but my guess is it should work.

Reply:
I tried this based on the example [here](http://fs2.io/concurrency-primitives.html) but it was slow-going so I'll let you handle this if you don't mind. I agree it seems cleaner that way. I look forward to studying your change.